feat(integrations): add litellm integration

constantinius opened the pull request:
Add a first implementation of the litellm integration, supporting completion and embeddings

Closes https://linear.app/getsentry/issue/PY-1828/add-agent-monitoring-support-for-litellm

sentrivana commented on sentry_sdk/integrations/litellm.py:
```diff
@@ -0,0 +1,281 @@
+    kwargs["_sentry_span"] = span
+
+    # Set basic data
+    set_data_normalized(span, SPANDATA.GEN_AI_SYSTEM, "litellm")
```
The actual provider as far as I understand. OTel says: 

> The Generative AI provider as identified by the client or server instrumentation.

The attribute itself is [deprecated](https://opentelemetry.io/docs/specs/semconv/registry/attributes/gen-ai/#gen-ai-system) and was renamed to [`gen_ai.provider.name`](https://opentelemetry.io/docs/specs/semconv/registry/attributes/gen-ai/#gen-ai-provider-name). I'm not sure what the plan regarding keeping up with OTel's semantic conventions is; you probably know better if we want to use the old or the new way.

constantinius replied:
We do have a summary issue for upgrading to [OTel 1.37](https://github.com/getsentry/sentry-python/issues/4753), so I'll keep the old one for now.

sentrivana commented on sentry_sdk/integrations/litellm.py:
```diff
@@ -0,0 +1,281 @@
+        )
+
+    # Record other parameters
+    params = {
+        "model": SPANDATA.GEN_AI_REQUEST_MODEL,
+        "stream": SPANDATA.GEN_AI_RESPONSE_STREAMING,
+        "max_tokens": SPANDATA.GEN_AI_REQUEST_MAX_TOKENS,
+        "presence_penalty": SPANDATA.GEN_AI_REQUEST_PRESENCE_PENALTY,
+        "frequency_penalty": SPANDATA.GEN_AI_REQUEST_FREQUENCY_PENALTY,
+        "temperature": SPANDATA.GEN_AI_REQUEST_TEMPERATURE,
+        "top_p": SPANDATA.GEN_AI_REQUEST_TOP_P,
+    }
```
Don't understand this comment, can you elaborate? What do the params have to do with `completion`?

constantinius replied:
`completion` takes quite generic `kwargs` that are then passed on to the model provider API. These above are used for `openai` (at least I suspect that this is where and how we retrieve them)

sentrivana commented on sentry_sdk/integrations/litellm.py:
```diff
@@ -0,0 +1,281 @@
+        if _input_callback not in litellm.input_callback:
+            litellm.input_callback.append(_input_callback)
+
+        litellm.success_callback = litellm.success_callback or []
+        if _success_callback not in litellm.success_callback:
+            litellm.success_callback.append(_success_callback)
+
+        litellm.failure_callback = litellm.failure_callback or []
+        if _failure_callback not in litellm.failure_callback:
+            litellm.failure_callback.append(_failure_callback)
```
There is definitely the potential for a timing issue but I don't see a way around it at the moment since the LiteLLM integration might not be in control of the overarching transaction.

From your testing when developing this, was this a real issue when something like a web framework was managing the transaction?

constantinius replied:
It was only an issue when writing code like this:

```python
with sentry_sdk.start_transaction(...):
    result = completion(...)
```

When using it in a framework (tried with FastAPI) I could not reproduce this error.

constantinius commented on sentry_sdk/integrations/litellm.py:
```diff
@@ -0,0 +1,281 @@
+    for key, value in litellm_params.items():
+        if value is not None:
+            set_data_normalized(span, f"gen_ai.litellm.{key}", value)
+
```
No, the callback is wrapped in `try`/`except` in litellm itself

constantinius commented:
> * Adding a test suite & the minimum supported version in `integrations/__init__.py`: see [`master`/scripts/populate_tox/README.md#add-a-new-test-suite](https://github.com/getsentry/sentry-python/blob/master/scripts/populate_tox/README.md?rgh-link-date=2025-10-02T08%3A17%3A24Z#add-a-new-test-suite) for more info
I've added tests and added the minimum versions to the list. I hope that works out.

> * If we want the SDK to have an `litellm` extra as mentioned in [the docs PR](https://github.com/getsentry/sentry-docs/pull/15091/files), it needs to be added to `setup.py`

Done

sentrivana reviewed (approved):
Took the liberty to add the missing test setup (the tests were not running in ci) and fix mypy's complaints, so if ci is green we should be good to go.
